### ISSUE TITLE
Numerical instability and NaN gradients in model_s due to manual sigmoid calculation

### ISSUE THREAD

**Issue by KCGallagher** [shahcompbio/scdna_replication_tools#15]
**Description:**

When running the PERT model on more complex scWGS datasets that require a higher number of training iterations, the ELBO loss smoothly decreases but suddenly collapses to NaN. This results in a CUDA assertion crash in the downstream Bernoulli distribution: `Assertion 0 <= p4 && p4 <= 1 failed` during Step 2 (S-phase inference). 

PyTorch's Anomaly Detector (`torch.autograd.set_detect_anomaly(True)`) attributes the failure to the `ExpBackward0` node at the phi probability calculation in `pert_model.py`: `Pythonphi = 1 / (1 + torch.exp(-a * t_diff))`. The current implementation calculates a manual sigmoid function followed by an in-place tensor mask to bound the probabilities: 
```Python
phi = 1 / (1 + torch.exp(-a * t_diff))
phi[phi<0.001] = 0.001
phi[phi>0.999] = 0.999
```

When -a * t_diff becomes a large positive number, torch.exp() overflows to inf. While the forward pass handles this (evaluating $1 / \infty$ as 0.0, which is then caught by the in-place mask), the backward pass attempts to calculate the derivative of $e^x$, resulting in a $0 \times \infty$ operation. This evaluates to NaN, which causes the downstream Bernoulli distribution to fail its [0, 1] boundaries. Furthermore, in-place masking (`phi[...] = ...`) is not safely tracked by PyTorch's autograd engine for gradient limitation. 

**Proposed Solution:**
Replace the manual exponential calculation and in-place masking with PyTorch's numerically stable torch.sigmoid() and torch.clamp(). This bypasses the exponential overflow in C++ and safely halts the gradients at the boundaries. I will open a PR with this fix shortly.
